src/QirTools/pyqir/qirlib/src/emit/qir/instructions.rs

```diff
@@ -0,0 +1,167 @@
+use super::{
+    array1d::{self, create_ctl_wrapper},
+    basic_values, calls,
+};
+use crate::{emit::Context, interop::Instruction};
+use inkwell::values::{BasicValueEnum, FunctionValue};
+use std::collections::BTreeMap;
+
+fn get_qubit<'ctx>(
+    name: &String,
+    qubits: &BTreeMap<String, BasicValueEnum<'ctx>>,
+) -> BasicValueEnum<'ctx> {
+    qubits.get(name).unwrap().to_owned()
+}
+
+fn get_register<'ctx>(
+    name: &String,
+    registers: &BTreeMap<String, (BasicValueEnum<'ctx>, Option<u64>)>,
+) -> (BasicValueEnum<'ctx>, Option<u64>) {
+    registers.get(name).unwrap().to_owned()
+}
+
+pub(crate) fn emit<'ctx>(
+    context: &Context<'ctx>,
+    inst: &Instruction,
+    qubits: &BTreeMap<String, BasicValueEnum<'ctx>>,
+    registers: &BTreeMap<String, (BasicValueEnum<'ctx>, Option<u64>)>,
+) -> () {
+    let intrinsics = &context.intrinsics;
+    let find_qubit = |name| get_qubit(name, qubits);
+    let ctl = |value| create_ctl_wrapper(context, value);
+    match inst {
+        Instruction::Cx(inst) => {
+            let control = ctl(&find_qubit(&inst.control));
+            let qubit = find_qubit(&inst.target);
+            controlled(context, intrinsics.x_ctl, control, qubit);
+        }
```

Comment:
Several of these look fairly similar, is it possible to consolidate the `Cx` and `Cz` arms, for example?